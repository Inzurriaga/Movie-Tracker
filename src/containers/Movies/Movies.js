import React, { Component } from "react";
import { Link } from "react-router-dom";
import { connect } from 'react-redux';
import { fetchMovies } from '../../api';
import PropTypes from 'prop-types';


export class Movies extends Component {
  constructor() {
    super();
    this.state = {
      moviesToSave: [],
      movieGenreTitles: [],
      pageTitle: '',
    }
  }

  componentDidMount = () => {
    if(this.props.id === "allMovies") {
      this.fetchGenresCategories()
    } else if (this.props.id === "inTheater"){
      this.fetchInTheater()
    } else if ( this.props.id === "upcoming"){
      this.fetchUpComing()
    } else if (this.props.id === "genres") {
      this.setAllMovies(this.props.genreInfo.results)
    } else if (this.props.id === "favorites") {
      this.fetchFavoriteMovies(this.props.favorites)
    }
  }

  fetchInTheater = async () => {
    const url = "https:api.themoviedb.org/3/discover/movie?"
    const endurl = "&with_release_type=2|3"
    const unresolvedTheater = await fetchMovies(url, endurl)
    this.setState({
      moviesToSave: unresolvedTheater.results
    })
  }

  fetchUpComing = async () => {
    const url = "http://api.themoviedb.org/3/movie/upcoming?"
    const endUrl = "&region=US"
    const unresolvedTheater = await fetchMovies(url, endUrl)
    this.setState({
      moviesToSave: unresolvedTheater.results
    })
  }

  fetchGenresCategories = async () => {
    const url = 'https://api.themoviedb.org/3/genre/movie/list?'
    try {
      const unresolvedTitles = await fetchMovies(url, '')
      const allGenres = await this.fetchAllGenres(unresolvedTitles.genres)
      const resolvedTitles = this.cleanGenreTitles(unresolvedTitles.genres)
      this.setState({
        movieGenreTitles: resolvedTitles
      })
      console.log(this.state);
    } catch (error) {
      console.log(error.message);
    }
  }

  cleanGenreTitles = (unresolvedTitles) => {
    return unresolvedTitles.map(title => {
      return title.name
    })
  }

  fetchAllGenres = async (genreArr) => {
    const url = "https:api.themoviedb.org/3/discover/movie?";
    let genreTitles = []
    try {
      const unresolvedGenres = genreArr.map( async (genre) => {
        const backEndUrl = `&with_genres=${genre.id}`;
        const singleGenre = await fetchMovies(url, backEndUrl)
        return singleGenre.results;
      })
      const allGenres = await Promise.all(unresolvedGenres)
      this.setAllGenres(allGenres)
    } catch(error) {
      console.log(error.message)
    }
  }

  setAllGenres = (allGenres) => {
    this.setState({
      moviesToSave: allGenres,
      pageTitle: '',
    })

  }

  setAllMovies = (allMovies) => {
    this.setState({
      moviesToSave: allMovies
    })
  }

  fetchFavoriteMovies = async (favArr) => {
    try {
      const unresolvedFavorites = favArr.map( async (movieId) => {
        let favUrl = `https://api.themoviedb.org/3/movie/${movieId}?`;
        const singleFavorite = await fetchMovies(favUrl, '')
        return singleFavorite
      })
      const allFavoriteMovies = await Promise.all(unresolvedFavorites)
      this.setFavorites(allFavoriteMovies)
    } catch(error) {
      console.log(error.message)
    }
  }

  setFavorites = (favoritesMovies) => {
    this.setState({
      moviesToSave: favoritesMovies
    })
  }

  render() {
      const { moviesToSave, movieGenreTitles } = this.state;
      let moviesToRender = null;
      if(!(this.props.id === "allMovies")) {
        moviesToRender = moviesToSave.map((movie, index) => {
            let backgroundImage = {backgroundImage: `url( http://image.tmdb.org/t/p/w1280${movie.backdrop_path})`}
            return (
                <Link className="genre-container" key={index} to={`/movies/${movie.id}`}>
                  <div className="genre-image" style={backgroundImage}>
                    <h3>{movie.title}</h3>
                    <p>{movie.overview}</p>
                  </div>
                </Link>
            )
        })
      } else if (this.props.id === "allMovies"){
        moviesToRender = moviesToSave.map((row, index) => {
          return (
            <div key={index} className="allMovies-column">
              <h4 className="Movies-SubTitle">{movieGenreTitles[index]} Movies</h4>
                <div className="allMovies-row">
                {
                  row.map(movie => {
                    let backgroundImage = {backgroundImage: `url( http://image.tmdb.org/t/p/w500${movie.backdrop_path})`}
                    return (
                        <Link className="allMovie-container" key={movie.id} to={`/movies/${movie.id}`}>
                          <div className="allMovie-image" style={backgroundImage}>
                            <h3>{movie.title}</h3>
                          </div>
                        </Link>
                    )
              })}
              </div>
            </div>
          )
        })
      }

      if(this.props.id === 'genres' || this.props.id === 'favorites' || this.props.id === 'upcoming' || this.props.id === 'inTheater') {
        return (
            <div className="Genre-Movies">
                <h2 className="Genre-Movies-Title">{this.props.title}</h2>
                <div className="Genre-Movies-Display">{moviesToRender}</div>
            </div>
        )
      } else {
        return (
            <div className="All-Movies">
                <h2 className="All-Movies-Title">{this.props.title}</h2>
                <div className="All-Movies-Display">{moviesToRender}</div>
            </div>
        )
      }

  }
}

Movies.propTypes = {
  	favorites: PropTypes.array,
    id: PropTypes.string,
    title: PropTypes.string,
}

export const mapStateToProps = (state) => ({
  favorites: state.favorites
})

export default connect(mapStateToProps)(Movies);
