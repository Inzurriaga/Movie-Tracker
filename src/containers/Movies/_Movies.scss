.Genre-Movies {
  width: 100%;
  position: relative;
  box-sizing: border-box;
  overflow-x: hidden;

  .Genre-Movies-Title {
    position: absolute;
    font-size: 9em;
    top: 15%;
    line-height: 0;
    left: 73.5%;
    text-transform: uppercase;
    color: #7b7b7b;
     mix-blend-mode: difference;
    letter-spacing: .05em;
    margin: 0;
    font-family: $font-Abel;
    z-index: 1;
    transform: rotate(90deg);
  }

  .Genre-Movies-Display {
    display: flex;
    justify-content: center;
    flex-wrap: wrap;
    position: relative;
    z-index: 9;
    box-sizing: border-box;
    width: 100%;
    padding: 2% 9% 2% 2%;

    .genre-container {
      display: block;
      text-decoration: none;
      width: 48%;
      padding: 1%;
      transition: all .5s ease;
      opacity: 1;
      cursor: pointer;

      &:hover {
        opacity: .5;
      }

      .genre-image {
        display: block;
        height: 100%;
        border-radius: 3px;
        position: relative;
        background-size: cover;
        background-position: top-center;
        background-repeat: no-repeat;
        display: flex;
        flex-direction: column;
        justify-content: flex-end;
        align-items: flex-end;
        height: 175px;
        box-shadow: 0px 2px 10px 3px #00000017;

        &:before {
          content: "";
          width: 100%;
          height: 100%;
          border-radius: 3px;
          position: absolute;
          background: linear-gradient(to bottom, rgba(0, 15, 255, 0) 0%, #755e0b5e 100%);
          top: 0px;
          padding: 1em;
          box-sizing: border-box;
          z-index: 2;
        }

        &:after {
          content: "";
          position: absolute;
          box-sizing: border-box;
          left: -15px;
          bottom: -65px;
          z-index: 1;
          transform: rotate(-45deg);
          border-top: 100px solid transparent;
          border-bottom: 100px solid transparent;
          border-right: 100px solid #231f20cf;
        }

        h3 {
          z-index: 9;
          margin: 0;
          width: 100%;
          padding: 0 .5em;
          color: #d4d4d4;
          font-family: $font-Sans;
          font-size: 2.5em;
          text-transform: unset;
          letter-spacing: .05em;
          box-sizing: border-box;
          text-shadow: 2px 2px 5px #231f20;
        }

        p {
          padding: 1em;
          z-index: 9;
          margin: 0;
          color: #fff;
          font-family: $font-copy;
          overflow: hidden;
          white-space: nowrap;
          text-overflow: ellipsis;
          max-width: 98.5%;
          font-size: .9em;
          font-weight: 100;
          letter-spacing: .03em;
          text-shadow: 2px 2px 5px #231f20;
          box-sizing: border-box;
        }
      }
    }
  }
}

.All-Movies {
  position: relative;
  padding: 2em;

  .All-Movies-Title {
    position: absolute;
    top: .6em;
    font-size: 9em;
    right: 0.1em;
    line-height: 0;
    text-transform: uppercase;
    color: #7b7b7b;
    mix-blend-mode: difference;
    letter-spacing: .05em;
    margin: 0;
    font-family: $font-Abel;
  }

  .All-Movies-Display {
    position: relative;
    display: flex;
    flex-direction: column;

    .allMovies-column {
      display: flex;
      flex-direction: column;

      h4.Movies-SubTitle {
        position: relative;
        color:#6f6f6f;
        font-family: $font-Sans;
        font-size: 1.2em;
        font-weight: 500;
        letter-spacing: .1em;
        text-transform: uppercase;
        margin: 0;
        line-height: 2;
        padding-top: 1em;
        mix-blend-mode: saturation;
      }

      .allMovies-row {
        overflow-x: scroll;
        width: 100%;
        white-space: nowrap;
        height: 150px;
        position: relative;

        &::after {
          content: '';
          position: absolute;
          right: 0px;
          top:0px;
          height: 100%;
          width: 50px;
          background: linear-gradient(to left, #231f20 0%, rgba(125, 185, 232, 0) 100%);
          z-index: 9;
        }

        a.allMovie-container {
          display: inline-block;
          width: 20%;
          margin-right: .4em;
          text-decoration: none;
          height: 100%;
          opacity: 1;
          transition: all .5s ease;

          &:hover {
            opacity: .7;
          }

          .allMovie-image {
            border-radius: 2px;
            background-size: cover;
            background-position: center;
            background-repeat: no-repeat;
            position: relative;
            width: 100%;
            height: 100%;
            display: flex;
            flex-direction: column;
            justify-content: flex-start;
            box-shadow: 0px 2px 10px 3px #00000017;

            &:before {
              content: "";
              width: 100%;
              height: 100%;
              border-radius: 3px;
              position: absolute;
              background: linear-gradient(to bottom, rgba(0, 15, 255, 0) 0%, #755e0b5e 100%);
              top: 0px;
              padding: 1em;
              box-sizing: border-box;
              z-index: 1;
            }

            &:after {
              content: "";
              width: 100%;
              height: 100%;
              border-radius: 3px;
              position: absolute;
              background: #67676726;
              top: 0px;
              padding: 1em;
              box-sizing: border-box;
              z-index: 1;
            }

            h3 {
              z-index: 9;
              margin: 0;
              width: 100%;
              padding: 1em;
              text-align: center;
              color: #d4d4d4;
              font-family: $font-Sans;
              white-space: normal;
              font-size: 1.25em;
              position: relative;
              top: 1em;
              letter-spacing: .05em;
              box-sizing: border-box;
              text-shadow: 2px 2px 5px #231f20;
            }
          }
        }
      }
    }

  }

}
