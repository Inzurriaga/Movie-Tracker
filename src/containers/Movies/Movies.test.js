import React from "react"
import { shallow } from "enzyme"
import { mapStateToProps, Movies } from "./Movies"

describe("Movies", () => {
    describe("Movies", () => {
        let wrapper;
        beforeEach(() => {
            wrapper = shallow(
                <Movies />
            )
        })
        it("should match snap shot", () => {
            expect(wrapper).toMatchSnapshot()
        })
        it("should have a default state", () => {
            expect(wrapper.state()).toEqual({
                moviesToSave: [],
                movieGenreTitles: [],
                pageTitle: '',
              })
        })
        it("should change the state of moviesToSave when setAllMovies is invoke", () => {
            expect(wrapper.state("moviesToSave")).toEqual([])
            const mockdata = [{name: "batman"}, {name: "superman"}]
            wrapper.instance().setAllMovies(mockdata)
            expect(wrapper.state("moviesToSave")).toEqual(mockdata)
        })
        it("", () => {
            expect(wrapper.state("moviesToSave")).toEqual([])
            const mockdata = [{genre: "horror"}, {genre: "action"}]
            wrapper.instance().setAllGenres(mockdata)
            expect(wrapper.state("moviesToSave")).toEqual(mockdata)
        })
    })
    describe("mapStateToProps", () => {
        it("should return an object", () => {
            const mockdata = {
                movies: [],
                user: {},
                favorites: []
            }
            const expected = {
                favorites: []
            }
            const result = mapStateToProps(mockdata)
            expect(result).toEqual(expected)
        })
    })
})