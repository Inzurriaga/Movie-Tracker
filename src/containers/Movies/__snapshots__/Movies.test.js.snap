// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movies Movies should match snap shot 1`] = `
<div
  className="All-Movies"
>
  <h2
    className="All-Movies-Title"
  />
  <div
    className="All-Movies-Display"
  />
</div>
`;
